### ISSUE TITLE
Parsing of ServerResponse "Body" broken in v0.0.1

### ISSUE THREAD

**Issue by Symmetric** [ClusterHQ/powerstrip#61]
I've been tracking down a pernicious issue that was preventing me from modifying the ServerResponse body on a post-hook. This is running 0.0.1, so this may well have been fixed on master.

I was constructing the body with code like this:

```python
body = json.loads(server_response["Body"])
# Do something to body...
server_response['Body'] = json.dumps(body, separators=(',', ':'))
```

If the command is `docker inspect`, then this fails with the error 
```
unexpected end of JSON input
[]
```

Looking at the pcap, the http content-length was being mis-calculated, and truncating the json data part way through.

I looked closely at the JSON of the working pass-through case, and saw this:

working:
```
{
  "PowerstripProtocolVersion": 1, 
  "ModifiedServerResponse": {
    "Body": "{\"AppArmorProfile\":\"\",\"Args\":[],\"Config\":{\"AttachStderr\":false,\"AttachStdin\":false,\"AttachStdout\":false,\"Cmd\":[\"/bin/bash\"],\"CpuShares\":0,\"Cpuset\":\"\",\"Domainname\":\"\",\"Entrypoint\":null,\"Env\":[\"CALICO_IP=192.168.184.147\",\
"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"],\"ExposedPorts\":null,\"Hostname\":\"6db7c065f1d1\",\"Image\":\"ubuntu\",\"MacAddress\":\"\",\"Memory\":0,\"MemorySwap\":0,\"NetworkDisabled\":false,\"OnBuild\":null,\"OpenStdin\":false,\"PortSpecs
\":null,\"StdinOnce\":false,\"Tty\":true,\"User\":\"\",\"Volumes\":null,\"WorkingDir\":\"\"},\"Created\":\"2015-02-20T02:19:53.488371274Z\",\"Driver\":\"aufs\",\"ExecDriver\":\"native-0.2\",\"HostConfig\":{\"Binds\":null,\"CapAdd\":null,\"CapDrop\":null,\"ContainerID
File\":\"\",\"Devices\":[],\"Dns\":null,\"DnsSearch\":null,\"ExtraHosts\":null,\"IpcMode\":\"\",\"Links\":null,\"LxcConf\":[],\"NetworkMode\":\"none\",\"PortBindings\":{},\"Privileged\":false,\"PublishAllPorts\":false,\"RestartPolicy\":{\"MaximumRetryCount\":0,\"Name
\":\"\"},\"SecurityOpt\":null,\"VolumesFrom\":null},\"HostnamePath\":\"/var/lib/docker/containers/6db7c065f1d15b9a737f522403729c177a30464867835686f85f694cc63972e0/hostname\",\"HostsPath\":\"/var/lib/docker/containers/6db7c065f1d15b9a737f522403729c177a30464867835686f8
5f694cc63972e0/hosts\",\"Id\":\"6db7c065f1d15b9a737f522403729c177a30464867835686f85f694cc63972e0\",\"Image\":\"5ba9dab47459d81c0037ca3836a368a4f8ce5050505ce89720e1fb8839ea048a\",\"MountLabel\":\"\",\"Name\":\"/focused_hawking\",\"NetworkSettings\":{\"Bridge\":\"\",\"
Gateway\":\"\",\"IPAddress\":\"\",\"IPPrefixLen\":0,\"MacAddress\":\"\",\"PortMapping\":null,\"Ports\":null},\"Path\":\"/bin/bash\",\"ProcessLabel\":\"\",\"ResolvConfPath\":\"/var/lib/docker/containers/6db7c065f1d15b9a737f522403729c177a30464867835686f85f694cc63972e0/
resolv.conf\",\"State\":{\"Error\":\"\",\"ExitCode\":0,\"FinishedAt\":\"0001-01-01T00:00:00Z\",\"OOMKilled\":false,\"Paused\":false,\"Pid\":40681,\"Restarting\":false,\"Running\":true,\"StartedAt\":\"2015-02-20T02:19:53.59737267Z\"},\"Volumes\":{},\"VolumesRW\":{}}\n
", 
    "Code": 200, 
    "ContentType": "application/json"
  }
}
```

broken:
```
{
  "PowerstripProtocolVersion": 1, 
  "ModifiedServerResponse": {
    "Body": "{\"AppArmorProfile\": \"\",\"HostsPath\": \"/var/lib/docker/containers/51866e3219da4249fd3887150cf30b1b068e08605cb73753426ea39d3a1cedc9/hosts\",\"Created\": \"2015-02-20T02:23:29.259825879Z\",\"Image\": \"5ba9dab47459d81c0037ca3836a368a4f8ce5050505ce8972
0e1fb8839ea048a\",\"Args\": [],\"Driver\": \"aufs\",\"HostConfig\": {\"CapDrop\": null,\"PortBindings\": {},\"NetworkMode\": \"none\",\"SecurityOpt\": null,\"Links\": null,\"IpcMode\": \"\",\"LxcConf\": [],\"ContainerIDFile\": \"\",\"Devices\": [],\"CapAdd\": null,\"
Binds\": null,\"RestartPolicy\": {\"MaximumRetryCount\": 0,\"Name\": \"\"},\"PublishAllPorts\": false,\"Dns\": null,\"ExtraHosts\": null,\"DnsSearch\": null,\"Privileged\": false,\"VolumesFrom\": null},\"MountLabel\": \"\",\"VolumesRW\": {},\"State\": {\"Pid\": 41442
,\"OOMKilled\": false,\"Paused\": false,\"Running\": true,\"FinishedAt\": \"0001-01-01T00:00:00Z\",\"Restarting\": false,\"Error\": \"\",\"StartedAt\": \"2015-02-20T02:23:29.386172648Z\",\"ExitCode\": 0},\"ExecDriver\": \"native-0.2\",\"ResolvConfPath\": \"/var/lib/d
ocker/containers/51866e3219da4249fd3887150cf30b1b068e08605cb73753426ea39d3a1cedc9/resolv.conf\",\"Volumes\": {},\"Path\": \"/bin/bash\",\"HostnamePath\": \"/var/lib/docker/containers/51866e3219da4249fd3887150cf30b1b068e08605cb73753426ea39d3a1cedc9/hostname\",\"Proces
sLabel\": \"\",\"Config\": {\"Env\": [\"CALICO_IP=192.168.184.147\",\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"],\"Hostname\": \"51866e3219da\",\"Entrypoint\": null,\"PortSpecs\": null,\"Memory\": 0,\"OnBuild\": null,\"OpenStdin\": false,\"M
acAddress\": \"\",\"Cpuset\": \"\",\"User\": \"\",\"CpuShares\": 0,\"AttachStdout\": false,\"NetworkDisabled\": false,\"WorkingDir\": \"\",\"Cmd\": [\"/bin/bash\"],\"StdinOnce\": false,\"AttachStdin\": false,\"MemorySwap\": 0,\"Volumes\": null,\"Tty\": true,\"AttachS
tderr\": false,\"Domainname\": \"\",\"Image\": \"ubuntu\",\"ExposedPorts\": null},\"Id\": \"51866e3219da4249fd3887150cf30b1b068e08605cb73753426ea39d3a1cedc9\",\"NetworkSettings\": {\"MacAddress\": \"\",\"Bridge\": \"\",\"PortMapping\": null,\"IPPrefixLen\": 0,\"IPAdd
ress\": \"192.168.184.147\",\"Gateway\": \"\",\"Ports\": null},\"Name\": \"/adoring_goodall\"}", 
    "Code": 200, 
    "ContentType": "application/json"
  }
}
```

The only difference I can spot was the trailing `}\n"` at the end of the Body in the working case.

Changing my code to 
```python
body = json.loads(server_response["Body"])
# Do something to body...
server_response['Body'] = json.dumps(body, separators=(',', ':')) + '\n'
```

has not resolved the issue, so I'm at a loss.

**Comment by Symmetric** [comment]
A bit more info on the content-length. Here's the header for the working case:

```
HTTP/1.1 200 OK
Date: Fri, 20 Feb 2015 02:51:05 GMT
Content-Length: 1893
Content-Type: application/json
Server: TwistedWeb/14.0.0

```

And the header for the not-working case:

```
HTTP/1.1 200 OK
Date: Fri, 20 Feb 2015 02:48:40 GMT
Content-Length: 1894
Content-Type: application/json
Server: TwistedWeb/14.0.0

```

This is suspicious since the responses from my plugin server have different lengths:

Working:
```
HTTP/1.1 200 OK
Date: Fri, 20 Feb 2015 02:51:05 GMT
Connection: close
Content-Type: text/html
Content-Length: 2242
Server: TwistedWeb/14.0.0

```

Broken:
```
HTTP/1.1 200 OK
Date: Fri, 20 Feb 2015 02:48:40 GMT
Connection: close
Content-Type: text/html
Content-Length: 2258
Server: TwistedWeb/14.0.0

```

**Comment by binocarlos** [comment]
Hey - thanks for the bug report...

Could you possibly try testing the same thing against the latest master - there is a fix that might help with this...

We've got an automated build so it "should" be a case of `docker run clusterhq/powerstrip:latest` rather than `docker run clusterhq/powerstrip:v0.0.1` (abbreviated obviously)

Thanks!

**Comment by Symmetric** [comment]
I tested this on a0aaf02f3cf6f1f272c23992e237f19c57428128, and no luck unfortunately.

Any suggestions for further diags? I can provide pcap if that's useful.

Cheers.

**Comment by lukemarsden** [comment]
Could you share a minimal reproducible test case? Failing that, pcap.
Thanks!

**Comment by Symmetric** [comment]
Ok, try this out:
```bash
git clone git@github.com:Symmetric/calico-docker.git
cd calico-docker
git checkout powerstrip-bug-repro
cd node/powerstrip_adapter
sudo docker build -t adapter .
cd ../powerstrip
sudo docker build -t powerstrip .
sudo docker run --name=adapter --net=host -v /tmp:/var/log/calico -d adapter 
sudo docker run --name=powerstrip --net=host -v /var/run/docker.sock:/var/run/docker.sock -d powerstrip
sleep 5
sudo DOCKER_HOST=127.0.0.1:2375 docker inspect adapter

```
My output for this is:


This is  within the context of a larger repo, but I've decoupled the problem to just the powerstrip adapter and powerstrip itself. Hopefully that's not too much hassle for you guys to work with.

The handler is in powerstrip-calico.py, starting at line 58. I'm modifying the body in the function defined at line 172 -- where I'm currently just setting body['NetworkSettings']['IPAddress'] = '1.2.3.4'.

The actual code path is very short, though a bit hacky since this is a prototype branch.

Logs in /var/log/calico/powerstrip-calico.py -- should be verbosely outputting the before and after json dumps.

I'm using --net=host to make it easier to capture all the traffic (snoop on lo).

Hopefully this is useful!

Cheers

**Comment by lukemarsden** [comment]
Fixed in #70, thanks for contributing, @symmetric!
